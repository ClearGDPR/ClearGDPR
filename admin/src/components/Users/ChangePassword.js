import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-form';
import TextInput from '../core/cards/dashboard/TextInput';
import Loader from '../core/cards/dashboard/Loader';

export class ChangePassword extends React.Component {
  static propTypes = {
    isLoading: PropTypes.bool,
    onSubmit: PropTypes.func,
    touched: PropTypes.object,
    errors: PropTypes.object
  };

  validatePassword(password) {
    // TODO: todo add proper validation
    if (!password) {
      return {
        error: 'Field required'
      };
    }
    if (password.length < 8) {
      return {
        error: 'Password must have min. 8 characters'
      };
    }
    return {
      success: null
    };
  }

  renderForm() {
    return (
      <React.Fragment>
        <TextInput
          label="New password"
          placeholder="*********"
          error={
            this.props.touched &&
            this.props.touched['newPassword'] &&
            this.props.errors &&
            this.props.errors['newPassword']
          }
          field="newPassword"
          validate={this.validatePassword}
        />
        <TextInput
          label="Repeat new password"
          placeholder="*********"
          error={
            this.props.touched &&
            this.props.touched['newPasswordRepeat'] &&
            this.props.errors &&
            this.props.errors['newPasswordRepeat']
          }
          field="newPasswordRepeat"
          validate={this.validatePassword}
        />
        <div>
          <input type="submit" className="btn" value="Save" />
        </div>
      </React.Fragment>
    );
  }

  render() {
    return (
      <form onSubmit={this.props.onSubmit}>
        {this.props.isLoading ? <Loader /> : this.renderForm()}
      </form>
    );
  }
}

const changePasswordForm = props => (
  <Form onSubmit={submittedValues => props.onSubmit && props.onSubmit(submittedValues)}>
    {formApi => (
      <ChangePassword
        onSubmit={formApi.submitForm}
        errors={formApi.errors}
        touched={formApi.touched}
        {...props}
      />
    )}
  </Form>
);

changePasswordForm.propTypes = {
  onSubmit: PropTypes.func
};

export default changePasswordForm;
