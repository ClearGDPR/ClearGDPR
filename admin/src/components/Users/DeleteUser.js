import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-responsive-modal';

import Loader from 'components/core/cards/dashboard/Loader';
import { PrimaryButton } from 'components/core/Common/Buttons/Buttons';

const DeleteUser = props => (
  <Modal open={props.isOpen} showCloseIcon={false} onClose={() => {}} center>
    {!props.isLoading ? (
      <React.Fragment>
        <p>Are you sure you want to delete this user?</p>
        <p>
          <PrimaryButton
            text="Cancel"
            onClick={e => {
              e.preventDefault();
              props.onCancel();
            }}
          />
          &nbsp;
          <PrimaryButton
            text="OK"
            onClick={e => {
              e.preventDefault();
              props.onConfirm();
            }}
          />
        </p>
      </React.Fragment>
    ) : (
      <Loader />
    )}
  </Modal>
);

DeleteUser.propTypes = {
  onConfirm: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  isOpen: PropTypes.bool,
  isLoading: PropTypes.bool
};

export default DeleteUser;
