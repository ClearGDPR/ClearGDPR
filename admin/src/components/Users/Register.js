import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'informed';
import TextInput from 'components/core/Common/Forms/TextInput';
import Loader from 'components/core/cards/dashboard/Loader';

export class Register extends React.Component {
  static propTypes = {
    isLoading: PropTypes.bool,
    onSubmit: PropTypes.func,
    validatePassword: PropTypes.func,
    touched: PropTypes.object,
    errors: PropTypes.object
  };

  renderForm() {
    return (
      <React.Fragment>
        <TextInput
          label="Username"
          placeholder="mick.bern"
          error={
            this.props.touched &&
            this.props.touched['username'] &&
            this.props.errors &&
            this.props.errors['username']
          }
          field="username"
        />
        <TextInput
          label="New password"
          placeholder="*********"
          type={'password'}
          error={
            this.props.touched &&
            this.props.touched['password'] &&
            this.props.errors &&
            this.props.errors['password']
          }
          field="password"
          validate={this.props.validatePassword}
        />
        <div>
          <input type="submit" className="btn" value="Register" />
        </div>
      </React.Fragment>
    );
  }

  render() {
    return (
      <form onSubmit={this.props.onSubmit}>
        {this.props.isLoading ? <Loader /> : this.renderForm()}
      </form>
    );
  }
}

const registerForm = props => (
  <Form onSubmit={submittedValues => props.onSubmit && props.onSubmit(submittedValues)}>
    {formApi => (
      <Register
        {...props}
        onSubmit={formApi.submitForm}
        errors={{ ...formApi.errors, ...props.errors }}
        touched={formApi.touched}
      />
    )}
  </Form>
);

registerForm.propTypes = {
  onSubmit: PropTypes.func,
  errors: PropTypes.object
};

export default registerForm;
