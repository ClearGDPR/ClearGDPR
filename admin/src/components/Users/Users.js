import React from 'react';
import PropTypes from 'prop-types';

import Card from '../core/cards/dashboard/Card';
import Loader from '../core/cards/dashboard/Loader';

const Users = ({
  users,
  onChangePasswordClick,
  onRegisterUserClick,
  onDeleteClick,
  isLoading,
  children
}) => {
  function onChangePasswordClickHandler(e, userId) {
    e.preventDefault();
    onChangePasswordClick && onChangePasswordClick(userId);
  }

  function onDeleteClickHandler(e, userId) {
    e.preventDefault();
    onDeleteClick && onDeleteClick(userId);
  }

  function onRegisterUserClickHandler(e) {
    e.preventDefault();
    onRegisterUserClick && onRegisterUserClick();
  }

  function renderUsers() {
    return users.map((user, idx) => (
      <tr key={idx}>
        <td data-label="ID">{user.id}</td>
        <td data-label="Username">{user.username}</td>
        <td data-label="Actions">
          <button className="ui-action btn" onClick={e => onChangePasswordClickHandler(e, user.id)}>
            Change password
          </button>
          <button className="ui-action btn" onClick={e => onDeleteClickHandler(e, user.id)}>
            Delete
          </button>
        </td>
      </tr>
    ));
  }

  return (
    <React.Fragment>
      <section className="cards">
        <div className="action-bar">
          <div className="text">
            <h4>Users</h4>
            <p>Manage users that can access the administrative panel</p>
          </div>
          <div className="spacer" />
          <button className="ui-action btn" onClick={onRegisterUserClickHandler}>
            + Register User
          </button>
        </div>
        <div className="row">
          <Card cols={8}>
            <div className="content">
              {!isLoading ? (
                <table className="responsive-table">
                  <thead>
                    <tr>
                      <th>ID</th>
                      <th>Username</th>
                      <th>Actions</th>
                    </tr>
                  </thead>
                  <tbody>{renderUsers()}</tbody>
                </table>
              ) : (
                <Loader />
              )}
            </div>
          </Card>
        </div>
      </section>
      {children}
    </React.Fragment>
  );
};

Users.propTypes = {
  users: PropTypes.arrayOf(PropTypes.object),
  onChangePasswordClick: PropTypes.func,
  onDeleteClick: PropTypes.func,
  onRegisterUserClick: PropTypes.func,
  isLoading: PropTypes.bool,
  children: PropTypes.node
};

export default Users;
