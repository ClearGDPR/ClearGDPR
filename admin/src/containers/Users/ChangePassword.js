import React from 'react';
import PropTypes from 'prop-types';

import ChangePassword from 'components/Users/ChangePassword';
import config from 'config';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import internalFetch from 'helpers/internal-fetch';
import { toast } from 'react-toastify';

export class ChangePasswordContainer extends React.Component {
  static propTypes = {
    userId: PropTypes.number.isRequired,
    closePanel: PropTypes.func
  };

  state = {
    isLoading: false,
    errors: {}
  };

  async changePassword(userId, password) {
    await internalFetch(`${config.API_URL}/api/management/users/${userId}/update-password`, {
      method: 'POST',
      body: JSON.stringify({
        password
      })
    }).catch(err => {
      this.setState({
        errors: {
          '': err.message
        }
      });
      return Promise.reject(err);
    });
  }

  startLoading() {
    this.setState({
      isLoading: true
    });
  }

  validatePassword(password) {
    // TODO: todo add proper validation
    if (!password) {
      return 'Field required';
    }
    if (password.length < 8) {
      return 'Password must have min. 8 characters';
    }
    return null;
  }

  stopLoading = () => {
    this.setState({
      isLoading: false
    });
  };

  onSubmit = data => {
    const { newPassword, newPasswordRepeat } = data;

    if (newPassword !== newPasswordRepeat) {
      this.setState({
        errors: {
          newPasswordRepeat: 'Passwords do not match!'
        }
      });
      return;
    }

    this.startLoading();

    return this.changePassword(this.props.userId, newPassword)
      .then(this.stopLoading)
      .then(res => {
        toast.success('User successfully registered');
        return res;
      })
      .then(() => this.props.closePanel && this.props.closePanel())
      .catch(this.stopLoading);
  };

  render() {
    return (
      <ChangePassword
        errors={this.state.errors}
        validatePassword={this.validatePassword}
        onSubmit={this.onSubmit}
        isLoading={this.state.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => <ChangePasswordContainer {...props} closePanel={closePanel} />}
  </PanelConsumer>
);
