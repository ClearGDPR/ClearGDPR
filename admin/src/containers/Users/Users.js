import React from 'react';
import PropTypes from 'prop-types';

import Users from '../../components/Users/Users';
import config from '../../config';
import session from '../../helpers/Session';
import { PanelConsumer } from '../MainLayout/PanelContext';
import Form from '../../components/core/cards/dashboard/Form';

class UsersContainer extends React.Component {
  static propTypes = {
    openPanel: PropTypes.func
  };

  state = {
    users: [],
    isLoading: true
  };

  async getUsers() {
    const response = await fetch(`${config.API_URL}/api/management/users/list`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${session.getToken()}`
      }
    });

    return await response.json();
  }

  componentDidMount() {
    this.getUsers()
      .then(users =>
        this.setState({
          users,
          isLoading: false
        })
      )
      .catch(e => {
        console.error(e);
        this.setState({
          isLoading: false
        });
      });
  }

  openChangePasswordForm() {
    this.props.openPanel(Form);
  }

  render() {
    return (
      <Users
        users={this.state.users}
        isLoading={this.state.isLoading}
        onChangePasswordClick={this.openChangePasswordForm.bind(this)}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ openPanel }) => <UsersContainer {...props} openPanel={openPanel} />}
  </PanelConsumer>
);
