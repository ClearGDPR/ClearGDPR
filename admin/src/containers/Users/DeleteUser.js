import React from 'react';
import PropTypes from 'prop-types';

import { UsersConsumer } from './UsersContext';
import DeleteUser from 'components/Users/DeleteUser';

export class DeleteUserContainer extends React.Component {
  static propTypes = {
    userId: PropTypes.number.isRequired,
    deleteUser: PropTypes.func.isRequired,
    onClose: PropTypes.func,
    isOpen: PropTypes.bool,
    isLoading: PropTypes.bool
  };

  deleteUser = () => {
    this.props.deleteUser(this.props.userId).then(this.props.onClose);
  };

  render() {
    return (
      <DeleteUser
        onConfirm={this.deleteUser}
        onCancel={this.props.onClose}
        isOpen={this.props.isOpen}
        isLoading={this.props.isLoading}
      />
    );
  }
}

export default props => (
  <UsersConsumer>
    {({ deleteUser, isLoading }) => (
      <DeleteUserContainer {...props} isLoading={isLoading} deleteUser={deleteUser} />
    )}
  </UsersConsumer>
);
