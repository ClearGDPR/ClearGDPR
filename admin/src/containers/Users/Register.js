import React from 'react';
import PropTypes from 'prop-types';

import Register from 'components/Users/Register';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import { UsersConsumer } from './UsersContext';

export class RegisterContainer extends React.Component {
  static propTypes = {
    registerUser: PropTypes.func.isRequired,
    closePanel: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired
  };

  state = {
    errors: {}
  };

  validatePassword(password) {
    // TODO: todo add proper validation
    if (!password) {
      return 'Field required';
    }
    if (password.length < 8) {
      return 'Password must have min. 8 characters';
    }
    return null;
  }

  onSubmit(data) {
    const { username, password } = data;

    this.props
      .registerUser(username, password)
      .then(() => this.props.closePanel())
      .catch(e =>
        this.setState({
          errors: {
            username: e.toString()
          }
        })
      );
  }

  render() {
    return (
      <Register
        errors={this.state.errors}
        validatePassword={this.validatePassword}
        onSubmit={this.onSubmit.bind(this)}
        isLoading={this.props.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => (
      <UsersConsumer>
        {({ registerUser, isLoading }) => (
          <RegisterContainer
            {...props}
            isLoading={isLoading}
            registerUser={registerUser}
            closePanel={closePanel}
          />
        )}
      </UsersConsumer>
    )}
  </PanelConsumer>
);
