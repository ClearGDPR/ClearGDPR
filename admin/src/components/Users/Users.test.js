import React from 'react';
import { shallow } from 'enzyme';
import Users from './Users';

const props = {
  users: [
    {
      id: 1,
      username: 'admin'
    },
    {
      id: 2,
      username: 'joe'
    },
    {
      id: 3,
      username: 'marc'
    },
    {
      id: 4,
      username: 'gina'
    }
  ]
};

describe('(Component) Users', () => {
  it('should render correctly', () => {
    const component = shallow(<Users {...props} />);
    expect(component).toMatchSnapshot();
  });
});
