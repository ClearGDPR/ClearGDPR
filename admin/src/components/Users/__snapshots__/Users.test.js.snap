// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Users should render when loading correctly 1`] = `
<React.Fragment>
  <section
    className="cards"
  >
    <ActionBar
      desc="Manage users that can access the administrative panel"
      title="Users"
    >
      <Component
        onClick={[Function]}
        text="+ Register user"
      />
    </ActionBar>
    <div
      className="row"
    >
      <Card
        cols={8}
      >
        <div
          className="content"
        >
          <Loader />
        </div>
      </Card>
    </div>
  </section>
</React.Fragment>
`;

exports[`(Component) Users should render when not loading correctly 1`] = `
<React.Fragment>
  <section
    className="cards"
  >
    <ActionBar
      desc="Manage users that can access the administrative panel"
      title="Users"
    >
      <Component
        onClick={[Function]}
        text="+ Register user"
      />
    </ActionBar>
    <div
      className="row"
    >
      <Card
        cols={8}
      >
        <div
          className="content"
        >
          <Table
            actions={
              Array [
                Object {
                  "label": "Change password",
                  "onClick": [Function],
                },
                Object {
                  "label": "Delete",
                  "onClick": [Function],
                },
              ]
            }
            columns={
              Object {
                "id": Object {
                  "title": "ID",
                },
                "username": Object {
                  "title": "Username",
                },
              }
            }
            rows={
              Array [
                Object {
                  "id": 1,
                  "username": "admin",
                },
                Object {
                  "id": 2,
                  "username": "joe",
                },
                Object {
                  "id": 3,
                  "username": "marc",
                },
                Object {
                  "id": 4,
                  "username": "gina",
                },
              ]
            }
          />
        </div>
      </Card>
    </div>
  </section>
</React.Fragment>
`;
